### PR TITLE
Add s5j platform support for Tickless Kernel

### PR DESCRIPTION
**Pull request by lokeshbv**
1. s5j arch support for Tickless kernel feature.

Tickless feature can be enabled by enabling below configs through menuconfig,

> Hardware Configuration > Chip Selection > S5J Peripheral Support > [*] MCT Global TIMER
CONFIG_S5J_G_TIMER=y
CONFIG_S5J_MCT=y

> Kernel Features > Clocks and Timers > [*] Support tick-less OS
CONFIG_SCHED_TICKLESS=y

> Kernel Features > Clocks and Timers > [*]   Tickless alarm
CONFIG_SCHED_TICKLESS_ALARM=y

> Kernel Features > Clocks and Timers > (100) System timer tick period (microseconds)
CONFIG_USEC_PER_TICK=100

2. adds device driver support for mct global timer.

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check Result:
os/arch/arm/src/s5j/s5j_mct.c:430: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('

**Comment by seinfra** [comment]
Code Rule Check OK.

File: os/arch/arm/src/s5j/s5j_mct.c [hunk 1]
```diff
@@ -427,7 +427,7 @@ void s5j_mct_disable(FAR struct s5j_mct_priv_s *priv)
 
 	/* Stop timer and disable interrupt */
 	value = mct_getreg32(priv,  S5J_MCT_L_TCON_OFFSET);
-	value &= ~( S5J_MCT_L_TCON_TIMER_START |  S5J_MCT_L_TCON_INT_START);
+	value &= ~(S5J_MCT_L_TCON_TIMER_START |  S5J_MCT_L_TCON_INT_START);
```
Comment (pradeepns85) [inline os/arch/arm/src/s5j/s5j_mct.c]: remove extra space after |

File: apps/shell/tash_usleep.c [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+/// @file   tash_usleep.c
+/// @brief  function to usleep
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include "tash_internal.h"
+
+/*
+ *	Tash command to sleep for micro or milli seconds
```
Comment (pradeepns85) [inline apps/shell/tash_usleep.c]: why to mention as milisec?

File: os/arch/arm/src/s5j/chip/s5jt200_mct.h [hunk 3]
```diff
@@ -59,23 +59,23 @@
...
-#define S5J_MCT_FRCNTO_OFFSET		0x0014
-#define S5J_MCT_TCON_OFFSET			0x0020
-#define S5J_MCT_INT_CSTAT_OFFSET	0x0030
-#define S5J_MCT_INT_ENB_OFFSET		0x0034
-#define S5J_MCT_WSTAT_OFFSET		0x0040
+#define S5J_MCT_L_TCNTB_OFFSET		0x0000
+#define S5J_MCT_L_TCNTO_OFFSET		0x0004
+#define S5J_MCT_L_ICNTB_OFFSET		0x0008
+#define S5J_MCT_L_ICNTO_OFFSET		0x000C
+#define S5J_MCT_L_FRCNTB_OFFSET		0x0010
```
Comment (pradeepns85) [inline os/arch/arm/src/s5j/chip/s5jt200_mct.h]: can we align it properly ?

File: os/arch/arm/src/s5j/chip/s5jt200_mct.h [hunk 4]
```diff
@@ -95,4 +95,31 @@
 #define S5J_MCT_INT_ENB_FRC_MASK	(1 << 1)
 #define S5J_MCT_INT_ENB_ICNT_MASK	(1 << 0)
 
+/* Global MCT */
+
+#define S5J_MCT_G_CNT_L				(0x100)
```
Comment (pradeepns85) [inline os/arch/arm/src/s5j/chip/s5jt200_mct.h]: same as above

File: os/arch/arm/src/s5j/s5j_mct.c [hunk 5]
```diff
@@ -166,21 +214,197 @@ static bool g_mct_initialized;
...
+ *
+ * Description:
+ *	Read all 64-bits of global free running counter.
+ *	Returns the number of cycles in the global counter.
+ *
+ ****************************************************************************/
+
+u64 s5j_mct_global_get_frc_cnt(FAR struct s5j_mct_priv_s *priv)
+{
+	u32 low, high;
```
Comment (pradeepns85) [inline os/arch/arm/src/s5j/s5j_mct.c]: use 2 separate lines to define

File: os/arch/arm/src/s5j/s5j_mct.c [hunk 6]
```diff
@@ -166,21 +214,197 @@ static bool g_mct_initialized;
...
+ * Name: s5j_mct_global_set_frc_cnt
+ *
+ * Description:
+ *	Set FRC counter value
+ *
+ ****************************************************************************/
+
+void s5j_mct_global_set_frc_cnt(FAR struct s5j_mct_priv_s *priv, u64 count)
+{
+	u32 low, high;
```
Comment (pradeepns85) [inline os/arch/arm/src/s5j/s5j_mct.c]: same as above. coding style

File: os/arch/arm/src/s5j/s5j_mct.c [hunk 7]
```diff
@@ -166,21 +214,197 @@ static bool g_mct_initialized;
...
+ *	Set global timer comparation value.
+ *	MCT will set interrupt if reach to timer count.
+ *	FRC in global timer is an up-counter.
+ *	After it starts, it increments until 64'hffff_ffff_ffff_ffff and returns to 0.
+ *
+ ****************************************************************************/
+
+void s5j_mct_global_set_comp_cnt(FAR struct s5j_mct_priv_s *priv, int mct_id, uint32_t gmct_offset, u64 count)
+{
+	u32 low, high;
```
Comment (pradeepns85) [inline os/arch/arm/src/s5j/s5j_mct.c]: same as abov

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by sunghan-chang** [comment]
@lokeshbv What is purpose of usleep command on tash?

**Comment by lokeshbv** [comment]
@sunghan-chang , In Tickless, timer may support very low granularity . So we can even get a delay in terms of milli seconds. To verify that we can use 'usleep'
For ex: to sleep for 1ms ---> usleep 1000

**Comment by sunghan-chang** [comment]
@lokeshbv If it is not helpful or not useful to user, even if it is much useful on testing tickless, I don't know it is good. I want to ask "Is it useful to user even though we support sleep?".

**Comment by sunghan-chang** [comment]
1. There is only sleep command on linux. There is no usleep on linux. 
2. If you need usleep on TASH indeed, let's make a new config for usleep command and set n as a default value. How do you think?

**Comment by lokeshbv** [comment]
@sunghan-chang, It's for testing tickless. I am also not sure about whether it is useful to user or not.
As you mentioned, yes Linux also don't support 'usleep' command. 
Please share your opinion whether to keep it or not.

File: os/arch/arm/src/s5j/Kconfig [hunk 8]
```diff
@@ -194,6 +194,15 @@ config S5J_I2S_MAXINFLIGHT
 config S5J_MCT
 	bool
 	default n
+	depends on S5J_HAVE_MCT
+	---help---
+		Multi Core Timer (MCT)
+
+config S5J_G_TIMER
```
[the remark this answers is not in the document]
Comment (lokeshbv) [reply os/arch/arm/src/s5j/Kconfig]: ok

**Comment by seinfra** [comment]
Code Rule Check OK.

File: os/arch/arm/src/s5j/s5j_mct.c [hunk 9]
```diff
@@ -130,9 +132,56 @@ static void s5j_mct_wait_wstat(uint32_t base, uint32_t bitmask)
 	putreg32(bitmask, reg);
 }
 
+/****************************************************************************
+ * Name: s5j_mct_global_get_comp_cnt
+ *
+ * Description:
+ *   Return current counter value of comparator
```
Comment (pradeepns85) [inline os/arch/arm/src/s5j/s5j_mct.c]: Can you add info about input parameters , any assumptions made and return type . etc to all new APIs?

Comment (lokeshbv) [reply os/arch/arm/src/s5j/s5j_mct.c]: done

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by pradeepns85** [comment]
@sunghan-chang @samirsahu1947  Can you please review it ?

File: apps/examples/tickless_test/tickless_test_main.c [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+static unsigned int thread;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static void *thread_func(void *arg)
+{
+	unsigned int duration = (unsigned int)arg;
+	unsigned int cur_thread = thread;
```
Comment (pradeepns85) [inline apps/examples/tickless_test/tickless_test_main.c]: i think it's better to use getpid() ? instead of believing the global variable "thread"

**Comment by sunghan-chang** [comment]
@lokeshbv @girishks2000 We need to get DS changes about s5j bsp changes from SamsungARTIK/tizenrt.
This can cause conflicts. Is it ok to us?

**Comment by seinfra** [comment]
Target : [d5a8a8127b59807010fd94faaa8f4ef6eaabfab7] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d5a8a8127b59807010fd94faaa8f4ef6eaabfab7] -

**Pull request closed without merging**